### DISCUSSION TITLE
Mismatch data point count on recording rules

### DISCUSSION THREAD

**Discussion by rrraditya** [prometheus/prometheus#9826] [Q&A]
**What did you do?**
1. have a job (JobA) in prometheus and set of recording rules based on metrics from JobA
2. both interval was set to 2m
3. query both metrics and count_over_time it for range of 30m or 6h or 7d or 30d.
```
count_over_time(metrics{labelA="X"}[30m]) or count_over_time(metrics{labelA="X"}[6h])
count_over_time(recmetrics{labelA="X"}[30m]) or count_over_time(recmetrics{labelA="X"}[6h])
```

**What did you expect to see?**
raw metrics count_over_time for 30m is 15 data points, for 6h is 180 data points.
recording metrics count_over_time should be the same, 15 and 180 data points respectively.

**What did you see instead? Under which circumstances?**
raw metrics count_over_time for 30m is 15 data points, for 6h is 180 data points.
recording metrics count_over_time has less than that, around 13-14 data points and 178-180 data points up and down. it's like getting a race condition.

**Additional information**
1. we're using recording rules to grab SLO data which will calculate 30d of data via recording rules.
2. Because we're calculating 30d of data, the gap from anomaly above will get bigger and it makes the 30d of data not valid and reliable.
3. Our current block data saved in local disk for 35d is around 9.2G.
4. It's the second time this anomaly happened in our environment.
5. Our temporary resolution is to restart the prometheus service, and all the data point count is getting proper count for both raw metrics and record metrics.

**Environment**

* System information:
```
Linux 3.10.0-1127.10.1.el7.x86_64 x86_64
```
* Prometheus version:
```
prometheus, version 2.19.1 (branch: HEAD, revision: eba3fdcbf0d378b66600281903e3aab515732b39)
  build user:       root@62700b3d0ef9
  build date:       20200618-16:35:26
  go version:       go1.14.4
```
* Prometheus configuration file:
```
global:
  scrape_interval:     2m
  evaluation_interval: 2m
rule_files:
- /etc/prometheus/rules/rules-*.yml
```

* Prometheus service file:
```
ExecStart=/usr/local/bin/prometheus \
        --config.file /etc/prometheus/prometheus.yml \
        --web.enable-lifecycle \
        --web.enable-admin-api \
        --web.listen-address=:9090 \
        --storage.tsdb.path /var/lib/prometheus/ \
        --web.console.templates=/etc/prometheus/consoles \
        --web.console.libraries=/etc/prometheus/console_libraries \
        --storage.tsdb.max-block-duration=2h \
        --storage.tsdb.min-block-duration=2h \
        --storage.tsdb.retention.time=35d \
        --query.timeout=1h \
        --log.level=info
```
If you need more information that i haven't shared above, please do ask.
I will gladly to share it to help analyze this anomaly.

Thank you.

Regards,
Reqi

**Comment by rrraditya** [comment]
hi All,

does any of you have any thoughts or insights on this use case?

Thank you.

**Comment by roidelapluie** [comment]
Thank you for your report, I would also expect the same as you, but could you provide us with the raw data (run `my_metric[30m]` in a *instant* query and count the points you have)? Thank you.

**Comment by rrraditya** [comment]
hi @roidelapluie 

as i mentioned before, our count of raw data was correct as expected.
for 2m scrape interval, the count of raw data for 30m is 15 on our side.

![raw data](https://user-images.githubusercontent.com/76177906/142667565-15450c3d-f8c6-4573-b4bf-4024758451ea.jpg)

so the problem is more towards to the count of recording rules data.
for 2m evaluation interval, the count of recording rules data for 30m is not always 15, sometimes it goes to 14 or 13.

![recording rules data](https://user-images.githubusercontent.com/76177906/142667578-12bfb476-981d-4aa5-872d-daa9ce578543.jpg)

Thank you.

**Reply by roidelapluie** [reply]
Oh, I think there might be some delay in ingesting the data? Do you expose data with timestamp? Or maybe your scrapes are "slow"?

You could try: `count_over_time(metrics{labelA="X"}[30m] offset 2m)` in your rule if that is the case.

**Comment by rrraditya** [comment]
hi @roidelapluie 

I don't think so that we got delay on scraping the data from the exporter or the evaluation delay.
could you explain more with expose data with timestamp? the timestamp that i might able to referring to is the timestamp of the data point which when prometheus scarpe/evaluate the data.

regarding the "slow" terms that you mentioned, yes indeed we have some rules running on our prometheus. But this issue was occured in lite expression rules.

![rules duration](https://user-images.githubusercontent.com/76177906/142970615-bf6a5f50-60fc-4da1-ae6c-3b8190bd5c3e.jpg)

As you can see, the rules group calc was pretty high, and especially it will spike every 2h during the prometheus block data write process. This is the rules that we used to calculate 30d of SLO data.

But the one that I refer this issue to is the rules group sli.
For the whole group of SLI rules, avg evaluation for 3d around 0.07s or 7ms
![rules sli avg](https://user-images.githubusercontent.com/76177906/142970991-7e3bfe35-d5f5-4027-a315-8b0834c2cac6.jpg)

I'm sure it still consider as a safe number, so I think the "slow" terms not affected to the SLI rules.
Yet, at that time, when we try to query a 30m datapoint, it not giving output of a complete 15 datapoints.

I'll lookup for the offset query and give update back to you.
Do you have any more insights that I can look up to?

Thank you.

Regards,
Reqi

**Comment by rrraditya** [comment]
hi @roidelapluie 

after I do some research about the offset modifier, so it means that you want to do workaround where we skip the latest datapoint (2m) and start the datapoint from minutes 32 to minutes 2, instead of minutes 30 to latest datapoint. cmiiw.

Unfortunately I can't validate this workaround at this time, because everything still working properly and as expected. So I'm pretty sure I will get same result whether i'm using offset or not.

maybe you have some insights on my previous comment? because i think this issue wasn't affected by the slow evaluation caused by calc rules group where it calculate 30 days of data for SLO.

Thank you.

**Reply by roidelapluie** [reply]
"Unfortunately I can't validate this workaround at this time, because everything still working properly and as expected"

What do you mean? Does the recording rule work now?

**Reply by rrraditya** [reply]
yes, as i mentioned on my thread submission, in additional information section, number 5.

`Our temporary resolution is to restart the prometheus service, and all the data point count is getting proper count for both raw metrics and record metrics.`

because this anomaly happened twice already, that's why i tried to open this to github discussion to get other point of view and insights from others that might facing same use case before.

**Reply by roidelapluie** [reply]
is the hostname of your Prometheus server changing across restarts?

**Reply by rrraditya** [reply]
hmm.. i just restart prometheus service, we didn't change the hostname upon restart.
why'd you ask that? is there any use case that you know that similar like this but with changing hostname scenario?
and what is the explanation or the relation between those 2 activity?

Please confirm.

Thank you.

**Reply by roidelapluie** [reply]
The time Prometheus records rules and scrape metrics is calculated based on its hostname. you could have a race between the calculation of the rule and the scrape of the metrics. This is why if you run in docker, it's best to have a "fixed" hostname, and if you have a HA Pair of prometheus servers, they should have different hostnames.

I don't know by heart but there is also one year of bugfixes from 2.19, you could also try to upgrade your server.

**Comment by isaitgirl** [comment]
@rrraditya ,

I've been observing discrepancies between raw metrics and recorded ones.
I have a metric that is collected each 15s and a recorded metric each 5 minutes that applies a increase(xx [5m]) on the raw metric.
When I compare raw metric and recorded metric at a common informed timestamp (applying increase(xx [5m] to metric VS the recorded metric) the values are different.

I'm trying to understand whether it has any relation to scrape_interval / evaluation_interval and a different data points count in each scenario (raw and recorded) or not. It's annoying, it's not precise.
